Sources/SwiftMemcache/MemcachedConnection.swift

```diff
@@ -0,0 +1,113 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-memcache-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+@_spi(AsyncChannel)
+
+import NIOCore
+import NIOPosix
+
+/// An actor to create a connection to a Memcache server.
+///
+/// This actor can be used to send commands to the server.
+public actor MemcachedConnection {
+    private typealias StreamElement = (MemcachedRequest, CheckedContinuation<MemcachedResponse, Error>)
+    /// The underlying channel to communicate with the server.
+    private let channel: NIOAsyncChannel<ByteBuffer, ByteBuffer>
+    /// The allocator used to create new buffers.
+    private let bufferAllocator: ByteBufferAllocator
+    /// The channel's event loop group.
+    private let eventLoopGroup: EventLoopGroup
+    /// The stream of requests to be sent to the server.
+    private let requestStream: AsyncStream<StreamElement>
+    /// The continuation for the request stream.
+    private let requestContinuation: AsyncStream<StreamElement>.Continuation
+
+    /// Initialize a new MemcachedConnection.
+    ///
+    /// - Parameters:
+    ///   - host: The host address of the Memcache server.
+    ///   - port: The port number of the Memcache server.
+    ///   - eventLoopGroup: The event loop group to use for this connection.
+    public init(host: String, port: Int, eventLoopGroup: EventLoopGroup) async throws {
+        self.eventLoopGroup = eventLoopGroup
+        let bootstrap = ClientBootstrap(group: self.eventLoopGroup)
+            .channelInitializer { channel in
+                return channel.pipeline.addHandlers([MessageToByteHandler(MemcachedRequestEncoder()), ByteToMessageHandler(MemcachedResponseDecoder())])
+            }
+
+        let rawChannel = try await bootstrap.connect(host: host, port: port).get()
+        self.channel = try NIOAsyncChannel<ByteBuffer, ByteBuffer>(synchronouslyWrapping: rawChannel)
+        self.bufferAllocator = ByteBufferAllocator()
+
+        let (stream, continuation) = AsyncStream<StreamElement>.makeStream()
+        self.requestStream = stream
+        self.requestContinuation = continuation
+    }
+
+    /// Start consuming the requestStream
+    ///
+    /// This function starts consuming the requests from the `requestStream`,
+    /// sending each request to the Memcache server and handling the server's responses.
+    public func run() async {
+        for await (request, continuation) in self.requestStream {
+            var buffer = self.bufferAllocator.buffer(capacity: 0)
+            let encoder = MemcachedRequestEncoder()
+            do {
+                try encoder.encode(data: request, out: &buffer)
+                try await self.channel.outboundWriter.write(buffer)
+
+                let responseBuffer = try await self.channel.inboundStream.first { _ in true }
```

Comment:
We still can't do this. Calling `first` is only going to allow us to do this once. We rather have to create an iterator at the start of the run method and then use that always. Something like this

```
var iterator = self.channel.inboundStream.makeAsyncIteartor()
...
let responseBuffer = try await iterator.next()
```

Reply:
correct! we can use the iterator to help consume and handle the request in order, it kinda ensures that the inbound stream remains open as long as the actors lifecycle, without getting terminated on the deinit of an iterator. we can avoid creating multiple iterators and having to encounter the actor-isolated mutation error.